### PR TITLE
Search - make wildcard field use constant scoring queries for wildcard queries

### LINKED ISSUE #69604: Wildcard field should use constant_score and reject wildcard query `rewrite` param
The wildcard field doesn't use a regular index that is useful for relevance ranking and should always return a Constant_score wrapper around any queries that produce non-constant scores.

As was noted in this [discuss post,](https://discuss.elastic.co/t/wildcard-query-rewrite-parameter-not-working-with-new-wildcard-field-type/265453/4) the rewrite parameter should also be rejected or ignored if anything other than constant_score.
Rejecting may be a little harsh if querying across indices that have a mix of keyword and wildcard field so ignoring the rewrite param on wildcard fields may be more appropriate. The docs should reflect whatever we decide is best (wildcard query or wildcard field docs? Or both?)

### PR DESCRIPTION
**Pull request by markharwood**
Wrap the wildcard queries produced by the wildcard field as a constant scoring query because there is no sensible index to use for scoring terms with their frequencies.
Added a note about ignoring rewrite parameters on wildcard queries.

Also includes fix where case insensitive/sensitive wildcard queries cached on same key. Run WIldcardFieldMapperTest with  seed `-Dtests.seed=722EAA9077BDA6AE` to reveal the issue.

Closes #69604

### CODE REVIEW

**Comment by markharwood** [comment]
Thanks to @jimczi for discovering the issue revealed as part of adding ConstantScoreQuery - there was a caching issue with AutomatonQueryOnBinaryDV objects due to not implementing hashcode/equals correctly for case sensitivity options. This PR also now includes a fix for that and related test addition.

File: x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/AutomatonQueryOnBinaryDv.java [hunk 1]
```diff
@@ -99,12 +96,13 @@ public boolean equals(Object obj) {
             return false;
           }
         AutomatonQueryOnBinaryDv other = (AutomatonQueryOnBinaryDv) obj;
-        return Objects.equals(field, other.field)  && Objects.equals(matchPattern, other.matchPattern);
+        return Objects.equals(field, other.field)  && Objects.equals(matchPattern, other.matchPattern)
+            && Objects.equals(automaton, other.automaton);
```
Comment (jimczi) [inline x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/AutomatonQueryOnBinaryDv.java:100]: `Automaton` does not implement equals/hashCode. The `ByteRunAutomaton` should be used instead.

File: x-pack/plugin/wildcard/src/test/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapperTests.java [hunk 2]
```diff
@@ -548,20 +550,47 @@ public void testWildcardAcceleration() throws IOException, ParseException {
...
+        assertNotEquals(csQ.hashCode(), ciQ.hashCode());
+
+        // Same query should be equal
+        Automaton caseSensitiveAutomaton2 = AutomatonQueries.toCaseInsensitiveWildcardAutomaton(
+            new Term("field", pattern),
+            Integer.MAX_VALUE
+        );
+        AutomatonQueryOnBinaryDv csQ2 = new AutomatonQueryOnBinaryDv("field", pattern, caseSensitiveAutomaton2);
+        assertNotEquals(csQ, csQ2);
+        assertNotEquals(csQ.hashCode(), csQ2.hashCode());
```
Comment (jimczi) [inline x-pack/plugin/wildcard/src/test/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapperTests.java:592]: Can you also test equality. There's a bug currently but this test doesn't catch it (automaton doesn't implement equals/hashCode).

File: x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java [hunk 3]
```diff
@@ -322,26 +322,22 @@ public Query wildcardQuery(String wildcardPattern, RewriteMethod method, boolean
...
                 verifyingBuilder.add(new BooleanClause(approxQuery, Occur.MUST));
                 verifyingBuilder.add(new BooleanClause(verifyingQuery, Occur.MUST));
-                return verifyingBuilder.build();
+                return new ConstantScoreQuery(verifyingBuilder.build());
             } else if (numWildcardChars == 0 || numWildcardStrings > 0) {
                 // We have no concrete characters and we're not a pure length query e.g. ???
                 return new DocValuesFieldExistsQuery(name());
             }
-            return verifyingQuery;
+            return new ConstantScoreQuery(verifyingQuery);
```
Comment (jimczi) [inline x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java:340]: The `AutomatonQueryOnBinaryDv` is already a constant score query so I don't think this is needed.

**Review by jimczi — CHANGES_REQUESTED** [review]
I left some comments.

**Comment by markharwood** [comment]
One thing I wasn't sure on was when the AutomatonQueryOnBinaryDV should convert the Automaton to a ByteRunAutomaton.
In the constructor or with every call to hashcode/equals/createWeight?

A ByteRunAutomaton looks to use a lot more memory than the source Automaton so might be expensive memory-wise to hold as a field. The alternative of converting Automaton to ByteRunAutomaton every time we want to use it might slow down comparisons in hash look-ups. I've gone with the latter but not sure if that's the better trade-off.

File: x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/AutomatonQueryOnBinaryDv.java [hunk 4]
```diff
@@ -99,12 +96,13 @@ public boolean equals(Object obj) {
             return false;
           }
         AutomatonQueryOnBinaryDv other = (AutomatonQueryOnBinaryDv) obj;
-        return Objects.equals(field, other.field)  && Objects.equals(matchPattern, other.matchPattern);
+        return Objects.equals(field, other.field)  && Objects.equals(matchPattern, other.matchPattern)
+            && Objects.equals(new ByteRunAutomaton(automaton), new ByteRunAutomaton(other.automaton));
```
Comment (jimczi) [inline x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/AutomatonQueryOnBinaryDv.java:100]: Rebuilding the `ByteRunAutomaton` on equals and hashCode seems overkill. Let's build it once in the ctr, the AutomatonQuery does the same so no reason to differ.

**Review by jimczi — APPROVED** [review]
I left one comment regarding the building of the BytesRunAutomaton. LGTM otherwise.

**Pull request merged**